docs/static/persistent-queues.asciidoc

```diff
@@ -1,44 +1,188 @@
+Apache Kafka). 
 
-In summary, the benefits of enabling persistent queues are as follows:
+NOTE: Persistent queues are disabled by default.
+To enable them, check out <<configuring-persistent-queues>>.
 
-* Absorbs bursts of events without needing an external buffering mechanism like
-Redis or Apache Kafka.
-* Provides an at-least-once delivery guarantee against message loss during
-a normal shutdown as well as when Logstash is terminated abnormally. If Logstash
-is restarted while events are in-flight, Logstash will attempt to deliver
-messages stored in the persistent queue until delivery succeeds at least once.
+[[persistent-queues-limitations]]
+==== Limitations of persistent queues
+
+Persistent queues do not solve these problems:
+
+* Input plugins that do not use a request-response protocol cannot be protected from data loss. Tcp, udp, zeromq push+pull, and many other inputs do not have a mechanism to acknowledge receipt to the sender. 
+(Plugins such as beats and http, which *do* have an acknowledgement capability, are well protected by this queue.)
+* A persistent queue does not handle permanent machine failures such as disk corruption, disk failure, and machine loss. 
+The data persisted to disk is not replicated.
+
+
+[[configuring-persistent-queues]]
+==== Configuring persistent queues
+
+To configure persistent queues, specify options in the Logstash <<logstash-settings-file,settings file>>:
+
+`queue.type`:: Specify `persisted` to enable persistent queues. By default, persistent queues are disabled (default: `queue.type: memory`).
+`path.queue`:: The directory path where the data files will be stored. By default, the files are stored in `path.data/queue`.
+`queue.type`:: Specify `persisted` to enable persistent queues. By default, persistent queues are disabled (default: `queue.type: memory`).
+`path.queue`:: The directory path where the data files will be stored. By default, the files are stored in `path.data/queue`. 
+`queue.page_capacity`:: The maximum size of a queue page in bytes. The queue data consists of append-only files called "pages". The default size is 64mb. Changing this value is unlikely to have performance benefits.
+`queue.drain`:: Specify `true` if you want Logstash to wait until the persistent queue is drained before shutting down. The amount of time it takes to drain the queue depends on the number of events that have accumulated in the queue. Therefore, you should avoid using this setting unless the queue, even when full, is relatively small and can be drained quickly. 
+// Technically, I know, this isn't "maximum number of events" it's really maximum number of events not yet read by the pipeline worker. We only use this for testing and users generally shouldn't be setting this.
+`queue.max_events`:: The maximum number of events that are allowed in the queue. The default is 0 (unlimited).
+`queue.max_bytes`:: The total capacity of the queue in number of bytes. The
```

Comment:
I think we should mention somewhere, that this is the value _per queue_, and that when using multiple pipelines configured with persistent queues, that the cumulative value of `queue.max_bytes` for each queue should be lower than the capacity of the user's disk. 

The confusion here is that this value can be set in `logstash.yml` and/or in `pipelines.yml`. If it is only set in `logstash.yml`, it will apply for _each pipeline_ that has a PQ enabled and is not overridden in `pipelines.yml`

Reply:
Can all? some? only this value be set in both logstash.yml and pipelines.yml?

Reply:
Thanks for bringing up this point. Will all _other_ queue settings in logstash.yml be overridden by the value in pipelines.yml?

Reply:
yes - we "document" this in the pipelines.yml we ship - https://github.com/elastic/logstash/blob/c698aa224b44c7356fbfeb58ebb0c4a66d72dc71/config/pipelines.yml#L38-L64